File: test-drive-xml-dom-xpath/src/mainjava/com/steelzack/test/drive/xml/XmlBookDomXpathParserManager.java
package com.steelzack.test.drive.xml;

import java.io.IOException;
import java.io.InputStream;
import java.util.Iterator;

import javax.xml.namespace.NamespaceContext;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpressionException;
import javax.xml.xpath.XPathFactory;

import org.w3c.dom.Document;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

public class XmlBookDomXpathParserManager extends XmlBookParserBuilder implements XmlBookParserManager {
	private static final String NAME = "@name";
	private static final String BOOKS = "/books/";
	private static final String BS = "bs:";
	private static final String AUTHOR = "author";
	private static final String GENRE = "genre";
	private static final String PAGES = "pages";
	private static final String BOOK = "book";
	private static final String HTTP_WWW_BOOKSHELFEXAMPLE_COM = "http://www.bookshelfexample.com";
	private Document document = null;
	private NodeList bookList = null;
	private XPath xpath = null;

	XmlBookDomXpathParserManager(InputStream ioStream) {
		super(ioStream);
	}

	@Override
	public String getBookTitle(int bookNumber) {
		try {
			return (String) xpath.evaluate(NAME, bookList.item(bookNumber), XPathConstants.STRING);
		} catch (XPathExpressionException e) {
			e.printStackTrace();
			return null;
		}
	}

	@Override
	public int getNumberOfPages(int bookNumber) {
		final String stringPages = getValueFromNode(bookNumber, PAGES);
		return stringPages == null ? -1 : Integer.parseInt(stringPages);
	}

	@Override
	public String getGenre(int bookNumber) {
		return getValueFromNode(bookNumber, GENRE);
	}

	@Override
	public String getAuthor(int bookNumber) {
		return getValueFromNode(bookNumber, AUTHOR);
	}

	@Override
	public int getNumberOfBooks() {
		return bookList.getLength();
	}

	@Override
	public void init() throws SAXException, IOException, ParserConfigurationException, XPathExpressionException {
		DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
		dbf.setNamespaceAware(true);
		DocumentBuilder db = dbf.newDocumentBuilder();
		document = db.parse(getIoStream());
		xpath = XPathFactory.newInstance().newXPath();
		xpath.setNamespaceContext(new NamespaceContext() {
			@Override
			public Iterator<String> getPrefixes(String namespaceURI) {
				return null;
			}

			@Override
			public String getPrefix(String namespaceURI) {
				return BS;
			}

			@Override
			public String getNamespaceURI(String prefix) {
				return HTTP_WWW_BOOKSHELFEXAMPLE_COM;
			}
		});
		bookList = (NodeList) xpath.evaluate(BOOKS + BS + BOOK, document, XPathConstants.NODESET);
	}

	private String getValueFromNode(int bookNumber, String nodeName) {
		try {
			return (String) xpath.evaluate(BS + nodeName, bookList.item(bookNumber), XPathConstants.STRING);
		} catch (XPathExpressionException e) {
			e.printStackTrace();
			return null;
		}
	}
}

File: jeorg-kotlin-apps/performance-parsers/xml-performance/test-drive-xml-dom-xpath/src/main/kotlin/org/jesperancinha/test/drive/xml/XmlBookDomXpathParserManager.kt
package org.jesperancinha.test.drive.xml

import org.w3c.dom.Document
import org.w3c.dom.NodeList
import org.xml.sax.SAXException
import java.io.IOException
import java.io.InputStream
import javax.xml.namespace.NamespaceContext
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException
import javax.xml.xpath.XPath
import javax.xml.xpath.XPathConstants
import javax.xml.xpath.XPathExpressionException
import javax.xml.xpath.XPathFactory

class XmlBookDomXpathParserManager(ioStream: InputStream) : XmlBookParserBuilder(ioStream) {
    private val document by lazy {
        val dbf = DocumentBuilderFactory.newInstance()
        dbf.isNamespaceAware = true
        val db = dbf.newDocumentBuilder()
        db.parse(ioStream)
    }
    val xpath: XPath by lazy {
        XPathFactory.newInstance().newXPath()
            .apply {
                namespaceContext = object : NamespaceContext {
                    override fun getPrefixes(namespaceURI: String): Iterator<String>? {
                        return null
                    }

                    override fun getPrefix(namespaceURI: String): String {
                        return BS
                    }

                    override fun getNamespaceURI(prefix: String): String {
                        return HTTP_WWW_BOOKSHELFEXAMPLE_COM
                    }
                }
            }
    }
    private val bookList: NodeList by lazy {
        xpath.evaluate(BOOKS + BS + BOOK, document, XPathConstants.NODESET) as NodeList
    }

    override fun getBookTitle(bookNumber: Int): String? {
        return try {
            xpath!!.evaluate(NAME, bookList!!.item(bookNumber), XPathConstants.STRING) as String
        } catch (e: XPathExpressionException) {
            e.printStackTrace()
            null
        }
    }

    override fun getNumberOfPages(bookNumber: Int): Int {
        val stringPages = getValueFromNode(bookNumber, PAGES)
        return stringPages?.toInt() ?: -1
    }

    override fun getGenre(bookNumber: Int): String? {
        return getValueFromNode(bookNumber, GENRE)
    }

    override fun getAuthor(bookNumber: Int): String? {
        return getValueFromNode(bookNumber, AUTHOR)
    }

    override val numberOfBooks: Int
        get() = bookList.length

    override fun init() {
    }

    @Throws(
        SAXException::class,
        IOException::class,
        ParserConfigurationException::class,
        XPathExpressionException::class
    )
    private fun getValueFromNode(bookNumber: Int, nodeName: String): String? {
        return try {
            xpath.evaluate(BS + nodeName, bookList.item(bookNumber), XPathConstants.STRING) as String
        } catch (e: XPathExpressionException) {
            e.printStackTrace()
            null
        }
    }

    companion object {
        private const val NAME = "@name"
        private const val BOOKS = "/books/"
        private const val BS = "bs:"
        private const val AUTHOR = "author"
        private const val GENRE = "genre"
        private const val PAGES = "pages"
        private const val BOOK = "book"
        private const val HTTP_WWW_BOOKSHELFEXAMPLE_COM = "http://www.bookshelfexample.com"
    }
}